confgenerator/confgenerator.go

```diff
@@ -97,6 +99,144 @@ func (m *Metrics) generateOtelPipelines() (map[string]otel.Pipeline, error) {
+	var out []fluentbit.Component
+	out = append(out, fluentbit.Service{}.Component())
+
+	if l != nil && l.Service != nil {
+		var sources []fbSource
+		var logNames []string
+		for pID, p := range l.Service.Pipelines {
+			for _, rID := range p.ReceiverIDs {
+				receiver, ok := l.Receivers[rID]
+				if !ok {
+					return nil, fmt.Errorf("receiver %q not found", rID)
+				}
+				tag := fmt.Sprintf("%s.%s", pID, rID)
+				components := receiver.Components(tag)
+				for i, pID := range p.ProcessorIDs {
+					processor, ok := l.Processors[pID]
+					if !ok {
+						return nil, fmt.Errorf("processor %q not found", pID)
+					}
+					components = append(components, processor.Components(tag, i)...)
+				}
+				components = append(components, setLogNameComponents(tag, rID)...)
+				logNames = append(logNames, regexp.QuoteMeta(rID))
+				sources = append(sources, fbSource{tag, components})
+			}
+		}
+		sort.Slice(sources, func(i, j int) bool { return sources[i].tag < sources[j].tag })
+
+		for _, s := range sources {
+			out = append(out, s.components...)
+		}
+		out = append(out, fluentbit.Stackdriver{
+			Match:     strings.Join(logNames, "|"),
+			Workers:   8,
+			UserAgent: userAgent,
+		}.Component())
+	}
+	// TODO: Use receivers instead of generating Tail objects directly.
+	out = append(out, fluentbit.Tail{
+		Tag:          "ops-agent-fluent-bit",
+		IncludePaths: []string{"${logs_dir}/logging-module.log"},
+	}.Component())
+	if hostInfo.OS != "windows" {
+		out = append(out, fluentbit.Tail{
+			Tag:          "ops-agent-collectd",
+			IncludePaths: []string{"${logs_dir}/metrics-module.log"},
+		}.Component())
+	}
+	out = append(out, fluentbit.Stackdriver{
+		Match:     "ops-agent-fluent-bit|ops-agent-collectd",
+		Workers:   8,
+		UserAgent: userAgent,
+	}.Component())
+
+	return out, nil
+}
+
+func setLogNameComponents(tag, logName string) []fluentbit.Component {
+	// TODO: Can we just set log_name_key in the output plugin and avoid this mess?
+	return []fluentbit.Component{
+		{
+			Kind: "FILTER",
+			Config: map[string]string{
+				"Match": tag,
+				"Add":   fmt.Sprintf("logName %s", logName),
+				"Name":  "modify",
+			},
+		},
+		{
+			Kind: "FILTER",
+			Config: map[string]string{
+				"Emitter_Mem_Buf_Limit": "10M",
+				"Emitter_Storage.type":  "filesystem",
+				"Match":                 tag,
+				"Name":                  "rewrite_tag",
+				"Rule":                  "$logName .* $logName false",
+			},
+		},
+		{
+			Kind: "FILTER",
+			Config: map[string]string{
+				"Match":  logName,
+				"Name":   "modify",
+				"Remove": "logName",
+			},
+		},
+	}
+}
+
+func (p LoggingProcessorParseShared) Components(tag string, i int) (fluentbit.Component, fluentbit.Component) {
+	parserName := fmt.Sprintf("%s.%d", tag, i)
+	filter := fluentbit.Component{
+		Kind: "FILTER",
+		Config: map[string]string{
+			"Match":    tag,
+			"Name":     "parser",
+			"Key_Name": p.Field,
+			"Parser":   parserName,
+		},
+	}
+	parser := fluentbit.Component{
```

Comment:
Might be worth commenting that the parser component returned from here is incomplete.

Reply:
Just a reminder that this would still be worth doing, IMO.

Reply:
This comment thread is still open. I plan to go back and comment every exported function after I delete all the unused code.